### DISCUSSION TITLE
USB redirection of web cam from Ubuntu client to Windows host

### DISCUSSION THREAD

**Discussion by dln949** [FreeRDP/FreeRDP#7024] [Q&A]
Hi, non-tech person here with a question.  I have a linux machine running freerdp so it is a client accessing a Windows 10 remote host.  This enables me to access the Windows 10 system from within my linux machine.  I am trying to redirect a usb webcam (that is connected to my linux machine) from my linux client into the Windows 10 machine so that it is accessible as I work within the Windows 10 display on my linux machine.  But, that camera displays nothing in Windows.  (Camera works fine on linux machine.)  This setup for the camera used to work a few months ago.  I use this command:  /usb:id,dev:19ff:0102      Version of xfreerdp is 2.2.0 from the Ubuntu repository.  

I don't know if the problem is with freerdp's usb redirect, or if the problem is within some setting in Windows.  I looked here on Github and saw several threads about this kind of problem, but after reading through them I am not able to figure out what is the current status:  Does usb redirect still work in freerdp?  Is there something in Windows that will stop this from working?  How can I determine which machine is the source of the problem?

(I did try asking this first on the Freenode IRC for Freerdp, and while I did get people responding, I was not able to get answers to my questions I am posing here.)

**Comment by thecmdrunner** [comment]
using `usbip` would work better if all you want to passthrough is a webcam.
Here's a video showing just that.
https://www.youtube.com/watch?v=gBCNLs_5pwM

**Comment by akallabeth** [comment]
@dln949 Sorry, was afk at that time :)
Currently there is only [MS-RDPUSB](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeusb/1a037956-736c-4e65-a39b-abb4642b8ad3)  redirection available, which is quite slow and hard to configure.

What you need to configure to have that working:
1. A windows version actually supporting this (I did not find a proper list which ones these are so far)
2. Your local user account needs access to `usb` devices (see [stackoverflow](https://stackoverflow.com/questions/40597515/libusb-calls-without-sudo-using-udev-rules) )
3. A USB device that is actually supported by this kind of redirection (stuff like mass storage, USB-HID, ... are not working at all due to server side not accepting these, multi configuration devices have issues too). Easiest is just try and have a look at where it fails

So, for you the following steps will (hopefully) yield a proper result:
1. Best start with latest `freerdp` release `2.3.2`
2. Start with `/log-filters:com.freerdp.channels.urbdrc.client:trace` to enable usb debug log messages
1. Check if the usb channel is working server side (either on the server or try to redirect and look for `[READ] RIM_EXCHANGE_CAPABILITY_REQUEST ...` messages
2. Check if the device is being redirected (some log messages with `ADD_VIRTUAL_CHANNEL`, if successful later on someting like `QUERY_DEVICE_TEXT`)

**Comment by dln949** [comment]
Thanks very much appreciate it.  

After considering this option and the earlier USBIP suggestion above, I came across this, "VirtualHere", which was very easy to get working for me:  

[https://www.youtube.com/watch?v=YsHzJHykxNA](url)

**Comment by oleg0421** [comment]
As of now, FreeRDP client already has support for [MS-RDPECAM] camera redirect protocol:
[Implement MS-RDPECAM client](https://github.com/FreeRDP/FreeRDP/pull/10258)
which is high level official RDP protocol extension, providing good performance for camera redirection.
